### ISSUE TITLE
Better support for shipping method providers (single method, multiple prices / options)

### ISSUE THREAD

**Issue by taxilian** [vendure-ecommerce/vendure#2649]
**Is your feature request related to a problem? Please describe.**

There are many shipping providers (e.g. shipstation, shippo, stamps.com, etc) which provide APIs which return an undetermined number of shipping methods with prices; currently in vendure a single Shipping Method can only return a single price, a single option.

**Describe the solution you'd like**
There should be some method for a "ShippingMethod" to return multiple options and some way when selecting shipping to set not only which ShippingMethod but which of the options it provided to use.

On discord (https://discord.com/channels/1100672177260478564/1200213503139655750/1200340563724083290) the suggestion was proposed to allow a shipping calculator to return multiple prices in an attempt to make it a non-breaking change.

**Describe alternatives you've considered**

Two options I've thought about:

1. Shipping plugin provides custom APIs for getting shipping methods, adds custom fields to the order for which is selected. The ShippingCalculator then looks at those fields to decide if it's going to be used.

2. The shipping calculator return the available shipping options as metadata; again custom fields can be used to set which method to use but if not provided it can return some arbitrary (or arbitrarily chosen?) value.

Both are a bit hacky, I feel like 2 is hackier than 1 =]

**Additional context**

Here is an abbreviated example of what the Shipstation APIs might return. (not exact, but it's what my custom API returns after getting all available carriers and services)

```json
[
    {
        name: 'Stamps.com',
        code: 'stamps_com',
        nickname: 'USPS / Stamps',
        shippingProviderId: 54321,
        primary: true,
        services: [
            {
                serviceName: 'USPS First Class Mail - Package',
                serviceCode: 'usps_first_class_mail',
                shipmentCost: 3.79,
                otherCost: 0,
            },
            {
                serviceName: 'USPS Priority Mail - Package',
                serviceCode: 'usps_priority_mail',
                shipmentCost: 6.64,
                otherCost: 0,
            },
        ],
    },
    {
        name: 'UPS by ShipStation',
        code: 'ups_walleted',
        nickname: 'UPS',
        shippingProviderId: 12345,
        primary: true,
        services: [
            {
                serviceName: 'UPS Next Day Air®',
                serviceCode: 'ups_next_day_air',
                shipmentCost: 17.76,
                otherCost: 0,
            },
            {
                serviceName: 'UPS® Ground',
                serviceCode: 'ups_ground',
                shipmentCost: 6.76,
                otherCost: 0,
            },
        ],
    },
]
```

**Comment by jacobfrantz1** [comment]
A somewhat related issue is error handling in shipping calculators. If an api call fails, it either has to throw which fails the entire getEligibleShippingMethods query, or return a fallback price; neither of which are great.

**Comment by brunoslalmeida** [comment]
Hi @michaelbromley ,

I'm diving into this feature request and looking forward to coding it up. Given its importance for my e-commerce platform, I wanted to double-check before starting my experiments. Are there any specific areas or functions I should be particularly careful not to modify?

**Comment by brunoslalmeida** [comment]
My current solution is to return from calculateFn until the GraphApi single result OR array of results. 

![Screenshot from 2024-02-16 21-12-04](https://github.com/vendure-ecommerce/vendure/assets/3662370/b9427b04-d5a1-4155-82a1-c4ad83da681e)
![Screenshot from 2024-02-16 21-12-46](https://github.com/vendure-ecommerce/vendure/assets/3662370/d134d9aa-4ad2-4976-911d-86d8117b1394)
![Screenshot from 2024-02-16 21-13-06](https://github.com/vendure-ecommerce/vendure/assets/3662370/944bb3af-3810-4056-86ce-07fea2ec7234)

And i've change all the other places that uses shipping result to if array take first (for tests only, not to brake so far).

My first question would be for the testing at admin-ui. The service packages/core/src/service/services/order-testing.service.ts can return an array, but I need help to understand what's the best way to allow graphql to return this TestShippingMethodResult | TestShippingMethodResult[]

**Comment by brunoslalmeida** [comment]
@michaelbromley i've added the code here [github](https://github.com/brunoslalmeida/vendure/tree/minor) it would be awesome to get some insights from you before moving forward.

**Comment by brunoslalmeida** [comment]
![Screenshot from 2024-02-19 00-15-38](https://github.com/vendure-ecommerce/vendure/assets/3662370/80fd7e49-e010-420b-b1be-8ea93d43b6bf)

**Comment by brunoslalmeida** [comment]
Yes, it is redundant it was already in plans to remove it. I was also struggling to find a name, quotes seams like a good name. I will fix it and continue.

**Comment by brunoslalmeida** [comment]
Hey @michaelbromley just finished you feedback changes (i was struggling with time) and will get the pace back this week. Thanks for your support.

**Comment by brunoslalmeida** [comment]
Hey @michaelbromley I have a question. 

As far as I am studding the entry-point for shipping selection is this mutation setOrderShippingMethod

**Comment by brunoslalmeida** [comment]
![Screenshot from 2024-02-26 21-49-23](https://github.com/vendure-ecommerce/vendure/assets/3662370/4930ec10-35ab-44c3-87af-9957a17740e6)

I am also testing with remix, and the front-end is considering the same item (probably due to the same method id). Is this a retro-compatibility problem?  Or if at the end a new version of the front end gets generated is not a problem ?

**Comment by michaelbromley** [comment]
Hi,

regarding the `setOrderShippingMethod` mutation, a non-breaking change could be to:

1. make the `shippingMethodId` input nullable (remove the `!`)
2. create a new input which takes an array of objects, whose shape you can decide based on the data you need.

Once we make changes to the API, we can update the storefront implementations as needed.

**Comment by kbanman** [comment]
I'm also looking for this functionality and can contribute to the implementation if there's a need, just let me know where I could be helpful.

I agree with the backward-compatible change suggested above. Don't know if there is an established pattern for deprecation but it would be good to discourage the use of the old signature after release so as to enable removing it at some point.

**Comment by kbanman** [comment]
@brunoslalmeida a small comment on your changes:

Instead of having `checkEligibilityByShippingMethod` return either a single or array, normalize the results to just returning an array. Since it's a private method we don't need to make it backward compatible. This saves on conditionals elsewhere.

Same goes for `ShippingMethod.apply`, simplify from `Promise<ShippingCalculationResult | ShippingCalculationResult[] | undefined>` to `Promise<ShippingCalculationResult[]>`

**Comment by brunoslalmeida** [comment]
Hey @kbanman we are facing a possible breaking change at other places. I've discuss with @michaelbromley at discord and will bring it here. The problem is that for on shippingMethod with multiple results there must be a new field (this is the db breaking) to select witch item from the results is being selected.

**Comment by brunoslalmeida** [comment]
This is the discussion thread. https://discord.com/channels/1100672177260478564/1218990088403288074

**Comment by brunoslalmeida** [comment]
> @brunoslalmeida a small comment on your changes:
> 
> Instead of having `checkEligibilityByShippingMethod` return either a single or array, normalize the results to just returning an array. Since it's a private method we don't need to make it backward compatible. This saves on conditionals elsewhere.
> 
> Same goes for `ShippingMethod.apply`, simplify from `Promise<ShippingCalculationResult | ShippingCalculationResult[] | undefined>` to `Promise<ShippingCalculationResult[]>`

@kbanman I do also think that retrocompatibility here might not be necessary. As soon as we handle te breaking problem with @michaelbromley I will test this change.

**Comment by michaelbromley** [comment]
So one way we could handle this in a non-breaking way would be to put it behind a feature flag like

```ts
shippingOptions: {
  multipleQuotesPerShippingMethod: true
}
```

and depending on that setting, we can dynamically define the field in the DB and extend the GraphQL API. In general I'm not a big fan of adding feature flags like this, because they complicate the code base in multiple places. But I'm not sure if there's a better solution without breaking things.

An alternative is to leave this for a future major release.

**Comment by brunoslalmeida** [comment]
The ideal solution would be to leave this for future major. I am currently planning to create a custom base plugin that allows developers to create their own solution with its supplier as a temporary solution while this is not solved via major. This custom base plugin will have a function base param so the developer can create it's own calls to the supplier and the plugin will handle the multiple shipping function using custom fields to add the selected method to the order.

**Comment by brunoslalmeida** [comment]
@michaelbromley I am moving back to this issue and would like your help with the feature flag and entity. I am a bit confused how to validade the feature flag at the entity file in order to just allow the new field creation if the parameter is true, can you give-me some light ?

**Comment by brunoslalmeida** [comment]
I am also thinking that it might be necessary to add validation if the calculator is returning an array and the multipleQuotesPerShippingMethod is false and notify user that this is problem. @michaelbromley Is this ok for you?

**Comment by michaelbromley** [comment]
Since entity properties (columns) are defined statically as class members with decorators, we'd need to take a different approach to add a property at run-time based on a flag. Luckily we do have a mechanism for this: [EntityMetadataModifier](https://docs.vendure.io/reference/typescript-api/configuration/entity-options/#entitymetadatamodifier).

So the idea would be to define a modifier function that adds the desired column to the entity during bootstrap. This happens [here](https://github.com/vendure-ecommerce/vendure/blob/2d2e51876e417a5b70779c8d5fb2880ca900b54b/packages/core/src/bootstrap.ts#L236) for user-defined modifiers. You could create one and run it directly after that in the case that the feature flag is set on.

Since this is adding a bit of complexity, try to think of ways to keep all flag-related code organized in a way that is clear and will be easy to refactor when we come to the next major and want to make it the default.

> I am also thinking that it might be necessary to add validation if the calculator is returning an array and the multipleQuotesPerShippingMethod is false and notify user that this is problem.

Yes that sounds reasonable.

**Comment by brunoslalmeida** [comment]
Ok, thanks I will take a look and do homework.

**Comment by brunoslalmeida** [comment]
@michaelbromley help clarify some doubts. 

[commit](https://github.com/brunoslalmeida/vendure/commit/10e086302b93ff215366bf83324a5aed2be37f71)

1. Adding metadataModifiers does not trigger the message to create migration, is this the correct behaviour or did I do something wrong? Running migration creates the field correctly.

2. Is at bootstrap the correct place to add the function ?

3. Do you have any other input ?

**Comment by michaelbromley** [comment]
> Adding metadataModifiers does not trigger the message to create migration, is this the correct behaviour or did I do something wrong? Running migration creates the field correctly.

The migration message "Your database schema does not match your current configuration..." is not actually part of the bootstrap process. It is shown in a standard installation because by default we run the `runMigrations` function prior to bootstrap which in turn calls [checkMigrationStatus](https://github.com/vendure-ecommerce/vendure/blob/7db3fbbcfc0f85513a730fc137a56b1c0c7a0b2a/packages/core/src/migrate.ts#L68)

> Is at bootstrap the correct place to add the function ?

Yes - the metadata modifier needs to be executed at that point.

> Any idea for the field name, I am using code just for tests now

This is a unique identifier for the given quote from a single ShippingMethod right? I think `code` is fine.

> Do you have any other input ?

This is out-of-scope for this specific issue, but the code makes me think we need a proper framework for implementing feature flags in a cleaner way. Right now we would have feature-flag code scattered around the code base. For one or two feature flags this is tolerable (if ugly) but the tendency is always towards more chaos :D so maybe it is a good time now to think about what the right way would be to handle this.

Imagine something like this: we define a `FeatureFlag` interface which allows us to make the process a bit more consistent:

```ts
abstract class FeatureFlag {
  name: string;
  abstract determineEnabledStatus(config: RuntimeVendureConfig): boolean;
  protected _isEnabled: boolean;

  get isEnabled() {
    return this._isEnabled;
  }
  
  constructor(config: RuntimeVendureConfig) {
    this._isEnabled = this.determineEnabledStatus(config);
  }
}
```

and then we define the flag like this:
```ts
// packages/core/src/feature-flags/multiple-shipping-quotes.ts
export class MultipleShippingQuotesFeatureFlag implements FeatureFlag {
  name: 'Multiple shipping quotes',
  determineEnabledStatus(config) => config.shippingOptions.multipleQuotesPerShippingMethod === true,
  
  // other logic/methods can be defined as methods 
  //on the feature flag class to better encapsulate things
  addShippingMethodCode(config: VendureConfig) {
    if (!config.entityOptions.metadataModifiers) {
      config.entityOptions.metadataModifiers = [];
    }

    const addShippingMethodCode: EntityMetadataModifier = metadata => {
      const instance = new ShippingLine();
      Column({ type: 'text', nullable: true })(instance, 'code');
    };
    config.entityOptions.metadataModifiers.push(addShippingMethodCode);
  }
}
```

At some point in the bootstrap process we instantiate all the feature flags:

```
// packages/core/src/feature-flags/index.ts
export const FEATURE_FLAG = {
  
};

// bootstrap.ts
FEATURE_FLAG.multipleShippingQuotes = new MultipleShippingQuotesFeatureFlag(config);
```

and use it like this:

```ts
// bootstrap.ts

if (FEATURE_FLAG.multipleShipping.isEnabled) {
  multipleShippingQuotesFeatureFlag.addShippingMethodCode(config);
}
```

Note: I don't expect you to implement this. I just wanted to jot down my idea as it came to me. For now you can continue development and it should be easy to refactor to use whatever feature flag system we implement in the future.

**Comment by dlhck** [comment]
Cross linking #3085 here, as feature flags have been discussed in this thread.

**Comment by cboothe** [comment]
Has anyone started working on this?

I believe the issue is that only one Shipping option is returned and it doesn't support an array of options.

**Comment by brunoslalmeida** [comment]
Hey @cboothe I have started, but after changing company I had to stop working on it due to compliance problems. Pleas let me know if you would like to continue from where I stopped. I can share the ideas I had at the time.

**Comment by cboothe** [comment]
Hey @brunoslalmeida 

Sure that would be great!

I am specifically interested in how you would change the return from shippingCalculator.

https://docs.vendure.io/reference/typescript-api/shipping/shipping-calculator#shippingcalculator

```
const flatRateCalculator = new ShippingCalculator({
  code: 'flat-rate-calculator',
  description: [{ languageCode: LanguageCode.en, value: 'Default Flat-Rate Shipping Calculator' }],
  args: {
    rate: {
      type: 'int',
      ui: { component: 'currency-form-input' },
    },
    taxRate: {
      type: 'int',
      ui: { component: 'number-form-input', suffix: '%' },
    },
  },
  calculate: (ctx, order, args) => {
    return {
      price: args.rate,
      taxRate: args.taxRate,
      priceIncludesTax: ctx.channel.pricesIncludeTax,
    };
  },
});
```

Returns

https://docs.vendure.io/reference/typescript-api/shipping/shipping-calculator#shippingcalculationresult

```
interface ShippingCalculationResult {
    price: number;
    priceIncludesTax: boolean;
    taxRate: number;
    metadata?: Record<string, any>;
}
```

How would **ShippingCalculationResult** support multiple values?

Something like this

```
"rates":[
      {
        "service_level":"Small Parcel",
        "service_name":"UPS Ground",
        "price":"23.01"
     },
    {
        "service_level":"UPS Three-Day Select",
        "service_name":"UPS Ground",
        "price":"67.73"
     },
    {
        "service_level":"Small Parcel",
        "service_name":"FedEx Ground",
        "price":"74.04"
     }
]
```

**Comment by brunoslalmeida** [comment]
```
const flatRateCalculator = new ShippingCalculator({
  code: 'flat-rate-calculator',
  description: [{ languageCode: LanguageCode.en, value: 'Default Flat-Rate Shipping Calculator' }],
  args: {
    rate: {
      type: 'int',
      ui: { component: 'currency-form-input' },
    },
    taxRate: {
      type: 'int',
      ui: { component: 'number-form-input', suffix: '%' },
    },
  },
  calculate: (ctx, order, args) => {
    return {
      price: args.rate,
      taxRate: args.taxRate,
      priceIncludesTax: ctx.channel.pricesIncludeTax,
    };
  },
});
```

The main idea is to return and array at calculate instead of a object:


```
type CalculateShippingFn<T extends ConfigArgs> = (
    ctx: RequestContext,
    order: Order,
    args: ConfigArgValues<T>,
    method: ShippingMethod,
) => CalculateShippingFnResult[]
```

**Comment by mezzat11** [comment]
Vote +1

**Comment by fobaz** [comment]
+1, this would be useful
